libsolidity/ast/Types.h

```diff
@@ -1454,6 +1455,54 @@ class TypeType: public Type
 	Type const* m_actualType;
 };
 
+/**
+ * The type of a UserDefinedValueType.
+ */
+class UserDefinedValueTypeType: public Type
```

Comment:
Why TypeType?
```
type NewType is int;
NewType a = NewType(2);
a;
```

In the above, the type of `a` (the one in isolation) is `UserDefinedValueType(int)`, while the type of `NewType` (as sub-expression of `NewType(2)` is `TypeType(UserDefinedValueType(int))`

If you treat the newly introduced type class more like the `Enum` or `Struct` class, then most of the other things should be automatic.